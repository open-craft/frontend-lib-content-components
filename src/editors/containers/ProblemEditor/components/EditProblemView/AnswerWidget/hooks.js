import { useState, useEffect } from 'react';
import { StrictDict } from '../../../../../utils';
import * as module from './hooks';
import { actions } from '../../../../../data/redux';

export const state = StrictDict({
  isFeedbackVisible: (val) => useState(val),
});

export const removeAnswer = ({ answer, dispatch }) => () => {
  dispatch(actions.problem.deleteAnswer({ id: answer.id }));
};

export const setAnswer = ({ answer, hasSingleAnswer, dispatch }) => (payload) => {
  dispatch(actions.problem.updateAnswer({ id: answer.id, hasSingleAnswer, ...payload }));
};

export const prepareFeedback = (answer) => {
  const [isFeedbackVisible, setIsFeedbackVisible] = module.state.isFeedbackVisible(false);
  useEffect(() => {
    // Show feedback fields if feedback is present
    const isVisible = !!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback;
    setIsFeedbackVisible(isVisible);
  }, [answer]);

  const toggleFeedback = (open) => {
    // Do not allow to hide if feedback is added
    if (!!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback) {
      setIsFeedbackVisible(true);
      return;
    }
    setIsFeedbackVisible(open);
  };
  return {
    isFeedbackVisible,
    toggleFeedback,
  };
};

export default {
  state, removeAnswer, setAnswer, prepareFeedback,
};
