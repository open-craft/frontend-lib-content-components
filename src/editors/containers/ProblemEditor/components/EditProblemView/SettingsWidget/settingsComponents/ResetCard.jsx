import React from 'react'
import { injectIntl, FormattedMessage, intlShape } from '@edx/frontend-platform/i18n';
import SettingsOption from '../SettingsOption';
import { Button, ButtonGroup, Form, Hyperlink } from '@edx/paragon';
import PropTypes from 'prop-types';
import { problemDataProps } from '../../../../../../data/services/cms/types';
import messages from '../messages';

export const ResetCard = ({
    showResetButton,
    updateSettings,
    intl,
}) => {

    return (
        <SettingsOption
            title={intl.formatMessage(messages.resetSettingsTitle)}
            summary={showResetButton ? intl.formatMessage(messages.resetSettingsTrue) : intl.formatMessage(messages.resetSettingsFalse)}
        >
            <div>
                <span>
                    <FormattedMessage {...messages.resetSettingText} />
                </span>
            </div>
            <div>
                <Hyperlink destination="#" target="_blank">
                    <FormattedMessage {...messages.advancedSettingsLinkText} />
                </Hyperlink>
            </div>
            <ButtonGroup size="lg" className="mb-2">
                <Button variant={showResetButton ? "outline-primary" : "primary"} onClick={() => updateSettings({ showResetButton: false })} >
                    <FormattedMessage {...messages.resetSettingsFalse} />
                </Button>
                <Button variant={showResetButton ? "primary" : "outline-primary"} onClick={() => updateSettings({ showResetButton: true })} >
                    <FormattedMessage {...messages.resetSettingsTrue} />
                </Button>
            </ButtonGroup>
        </SettingsOption>
    )
}

ResetCard.propTypes = {
    showResetButton: problemDataProps.settings.timeBetween,
    updateSettings: PropTypes.func.isRequired,
    intl: intlShape.isRequired,
};

export default injectIntl(ResetCard);