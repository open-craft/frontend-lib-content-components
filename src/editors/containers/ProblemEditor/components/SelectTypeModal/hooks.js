import {
  useState,
} from 'react';
import { StrictDict } from '../../../../utils';

export const state = StrictDict({
  selected: (val) => useState(val),
});

export default { state };
