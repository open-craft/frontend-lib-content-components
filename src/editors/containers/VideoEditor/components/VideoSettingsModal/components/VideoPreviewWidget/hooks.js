import messages from '../messages';
import { parseYoutubeId } from '../../../../../../data/services/cms/api';

function getVideoType(videoSource) {
  if (parseYoutubeId(videoSource) !== null) {
    return messages.videoTypeYoutube;
  }
  return messages.videoTypeOther;
}

export default {
  getVideoType,
};
