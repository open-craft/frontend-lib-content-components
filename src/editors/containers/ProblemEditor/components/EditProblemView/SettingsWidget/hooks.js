import { useState, useEffect } from 'react';

import * as module from './hooks';
import messages from './messages';
import { actions } from '../../../../../data/redux';
import _ from 'lodash-es';

export const state = {
    showAdvanced: (val) => useState(val),
    cardCollapsed: (val) => useState(val),
    summary: (val) => useState(val),
};

export const showAdvancedSettingsCards = () => {
    const [isAdvancedCardsVisible, setIsAdvancedCardsVisible] = module.state.showAdvanced(false);
    return {
        isAdvancedCardsVisible,
        showAdvancedCards: () => setIsAdvancedCardsVisible(true),
    };
}

export const showFullCard = () => {
    const [isCardCollapsed, setIsCardCollapsed] = module.state.cardCollapsed(false);
    return {
        isCardCollapsed,
        toggleCardCollapse: () => setIsCardCollapsed(!isCardCollapsed),
    };
}

export const hintsCardHooks = (hints, updateSettings) => {
    const [summary, setSummary] = module.state.summary({ message: messages.noHintSummary, values: {} })

    useEffect(() => {
        const hintsNumber = hints.length;
        if (hintsNumber == 0) {
            setSummary({ message: messages.noHintSummary, values: {} });
        } else {
            setSummary({ message: messages.hintSummary, values: { hint: hints[0].value, count: (hintsNumber - 1) } });
        }
    }, [hints]);

    const handleAdd = () => {
        let newId = 0
        if (!_.isEmpty(hints)){
            newId = Math.max(...hints.map(hint => hint.id)) + 1
        }
        const hint = { id: newId, value: "" }
        hints = [...hints, hint]
        updateSettings({ hints });
    }


    return {
        summary,
        handleAdd,
    }
}

export const hintsRowHooks = (id, hints, updateSettings) => {

    const handleChange = (event) => {
        const value = event.target.value
        hints = hints.map(hint => {
            if (hint.id === id) {
                return { ...hint, value };
            }
            return hint;
        });
        updateSettings({ hints });
    }

    const handleDelete = () => {
        hints = hints.filter((hint) => (hint.id != id));
        updateSettings({ hints });
    }

    return {
        handleChange,
        handleDelete,
    }
}

export const matlabCardHooks = (matLabApiKey, updateSettings) => {
    const [summary, setSummary] = module.state.summary({ message: "", values: {}, intl: false })

    useEffect(() => {
        if (_.isEmpty(matLabApiKey)) {
            setSummary({ message: messages.matlabNoKeySummary, values: {}, intl: true });
        } else {
            setSummary({ message: matLabApiKey, values: {}, intl: false });
        }
    }, [matLabApiKey])

    const handleChange = (event) => {
        updateSettings({ matLabApiKey: event.target.value });
    }

    return {
        summary,
        handleChange,
    }
}

export const randomizationCardHooks = (dispatch) => ({
    handleChange: (event) => {
        dispatch(actions.problem.updateSettings({ randomization: event.target.value }));
    }
})

export const resetCardHooks = (updateSettings) => {

    const setReset = (value) => {
        updateSettings({ showResetButton: value });
    }

    return {
        setResetTrue: () => setReset(true),
        setResetFalse: () => setReset(false),
    }
}

export const scoringCardHooks = (scoring, updateSettings) => {

    const handleMaxAttemptChange = (event) => {
        let unlimitedAttempts = true;
        let attemptNumber = parseInt(event.target.value);
        if(_.isNaN(attemptNumber)) {
            attemptNumber = 0;
        }
        if (attemptNumber > 0) {
            unlimitedAttempts = false;
        }
        updateSettings({ scoring: { ...scoring, attempts: { number: attemptNumber, unlimited: unlimitedAttempts } } });
    }

    const handleWeightChange = (event) => {
        let weight = parseFloat(event.target.value)
        if(_.isNaN(weight)) {
            weight = 0;
        }
        updateSettings({ scoring: { ...scoring, weight: weight } });
    }

    return {
        handleMaxAttemptChange,
        handleWeightChange,
    }
}

export const showAnswerCardHooks = (showAnswer, updateSettings) => {

    const handleShowAnswerChange = (event) => {
        updateSettings({ showAnswer: { ...showAnswer, on: event.target.value } });
    }

    const handleAttemptsChange = (event) => {
        let attempts = parseInt(event.target.value)
        if(_.isNaN(attempts)){
            attempts = 0;
        }
        updateSettings({ showAnswer: { ...showAnswer, afterAttempts: attempts } });
    }

    return {
        handleShowAnswerChange,
        handleAttemptsChange,
    }
}

export const timerCardHooks = (updateSettings) => ({

    handleChange: (event) => {
        let time = parseInt(event.target.value);
        if(_.isNaN(time)){
            time = 0;
        }
        updateSettings({ timeBetween: time });
    }
})

export const typeRowHooks = (typeKey, updateField) => {

    const onClick = () => {
        updateField({ problemType: typeKey });
    }

    return {
        onClick,
    }
}
