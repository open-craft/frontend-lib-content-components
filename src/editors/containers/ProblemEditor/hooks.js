import { useRef, useCallback, useState, useEffect } from 'react';
import { StrictDict } from '../../utils';
import * as module from './hooks';

export const state = StrictDict({
    refReady: (val) => useState(val),
  });

export const problemEditorConfig = ({
    setEditorRef,
    editorRef,
    question,
    updateQuestion
  }) => ({
    onInit: (evt, editor) => {
      setEditorRef(editor);
    },
    initialValue: question ? question : '',
    onFocusOut: () =>{
      let content = editorRef.current.getContent({format: 'text'});
      updateQuestion(content);
    },
  });

  export const prepareEditorRef = () => {
    const editorRef = useRef(null);
    const setEditorRef = useCallback((ref) => {
      editorRef.current = ref;
    }, []);
    const [refReady, setRefReady] = module.state.refReady(false);
    useEffect(() => setRefReady(true), []);
    return { editorRef, refReady, setEditorRef };
  };
