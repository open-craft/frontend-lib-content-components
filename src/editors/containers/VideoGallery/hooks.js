import React from 'react';
import { useSelector } from 'react-redux';
import * as module from './hooks';
import messages from './messages';
import * as appHooks from '../../hooks';
import { selectors } from '../../data/redux';
import analyticsEvt from '../../data/constants/analyticsEvt';
import {
  filterKeys,
  filterMessages,
  sortKeys,
  sortMessages,
  sortFunctions,
} from './utils';

export const {
  navigateCallback,
  navigateTo,
} = appHooks;

export const useSearchAndSortProps = () => {
  // const [searchString, setSearchString] = React.useState('');
  // const [sortBy, setSortBy] = React.useState(sortKeys.dateNewest);
  // const [filterBy, setFilterBy] = React.useState([]);
  const [hideSelectedVideos, setHideSelectedVideos] = React.useState(false);

  return {
    // searchString,
    // onSearchChange: (e) => setSearchString(e.target.value),
    // clearSearchString: () => setSearchString(''),
    // sortBy,
    // onSortClick: (key) => () => setSortBy(key),
    sortKeys,
    sortMessages,
    // filterBy,
    // onFilterClick: (keys) => setFilterBy(keys),
    filterKeys,
    filterMessages,
    showSwitch: false,
    hideSelectedVideos,
    switchMessage: messages.hideSelectedCourseVideosSwitchLabel,
    onSwitchClick: () => setHideSelectedVideos(!hideSelectedVideos),
  };
};

export const filterListBySearch = ({ searchString, videoList }) => (
  videoList.filter(({ displayName }) => displayName.toLowerCase().includes(searchString.toLowerCase()))
);

export const filterListByStatus = ({ statusFilter, videoList }) => {
  if (statusFilter.length < 1) {
    return videoList;
  }
  return videoList.filter(
    ({ status }) => statusFilter.map(key => filterKeys[key]).includes(status)
  );
};

export const filterListByHideSelectedCourse = ({ videoList }) => (
  // TODO Missing to implement this
  videoList
);

export const filterList = ({
  sortBy,
  filterBy,
  searchString,
  videos,
}) => {
  let filteredList = module.filterListBySearch({
    searchString,
    videoList: videos,
  });
  filteredList = module.filterListByStatus({
    statusFilter: filterBy,
    videoList: filteredList,
  });
  filteredList = module.filterListByHideSelectedCourse({
    videoList: filteredList,
  });
  return filteredList.sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);
};

export const useVideoListProps = ({ searchSortProps, videos }) => {
  const [highlighted, setHighlighted] = React.useState(null);
  const [
    showSelectVideoError,
    setShowSelectVideoError,
  ] = React.useState(false);
  const [
    showSizeError,
    setShowSizeError,
  ] = React.useState(false);
  // const filteredList = module.filterList({ ...searchSortProps, videos });
  const learningContextId = useSelector(selectors.app.learningContextId);
  const blockId = useSelector(selectors.app.blockId);
  return {
    galleryError: {
      show: showSelectVideoError,
      set: () => setShowSelectVideoError(true),
      dismiss: () => setShowSelectVideoError(false),
      message: messages.selectVideoError,
    },
    // TODO We need to update this message when implementing the video upload screen
    inputError: {
      show: showSizeError,
      set: () => setShowSizeError(true),
      dismiss: () => setShowSelectVideoError(false),
      message: messages.fileSizeError,
    },
    galleryProps: {
      // galleryIsEmpty: Object.keys(filteredList).length === 0,
      // searchIsEmpty: filteredList.length === 0,
      // displayList: filteredList,
      highlighted,
      onHighlightChange: (e) => setHighlighted(e.target.value),
      emptyGalleryLabel: messages.emptyGalleryLabel,
      showIdsOnCards: true,
      height: '100%',
    },
    selectBtnProps: {
      onClick: () => {
        if (highlighted) {
          navigateTo(`/course/${learningContextId}/editor/video/${blockId}?selectedVideoId=${highlighted}`);
        } else {
          setShowSelectVideoError(true);
        }
      },
    },
  };
};

export const useVideoUploadHandler = () => {
  const learningContextId = useSelector(selectors.app.learningContextId);
  const blockId = useSelector(selectors.app.blockId);
  return () => navigateTo(`/course/${learningContextId}/editor/video_upload/${blockId}`);
};

export const useCancelHandler = () => (
  navigateCallback({
    destination: useSelector(selectors.app.returnUrl),
    analytics: useSelector(selectors.app.analytics),
    analyticsEvent: analyticsEvt.videoGalleryCancelClick,
  })
);

export const useVideoProps = ({ videos }) => {
  const searchSortProps = useSearchAndSortProps();
  const videoList = useVideoListProps({ searchSortProps, videos });
  const {
    galleryError,
    galleryProps,
    inputError,
    selectBtnProps,
  } = videoList;
  const fileInput = { click: useVideoUploadHandler() };

  return {
    galleryError,
    inputError,
    fileInput,
    galleryProps,
    searchSortProps,
    selectBtnProps,
  };
};

export default {
  useVideoProps,
  useCancelHandler,
  useVideoUploadHandler,
};
