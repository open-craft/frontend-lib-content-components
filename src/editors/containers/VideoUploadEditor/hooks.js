import React from 'react';
import * as module from './hooks';
import { selectors } from '../../data/redux';
import store from '../../data/store';
import * as appHooks from '../../hooks';

const extToMime = {
  mp4: 'video/mp4',
  mov: 'video/quicktime',
};

export const {
  navigateTo,
} = appHooks;

export const state = {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  loading: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  errorMessage: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  textInputValue: (val) => React.useState(val),
};

export const uploadEditor = () => {
  const [loading, setLoading] = module.state.loading(false);
  const [errorMessage, setErrorMessage] = module.state.errorMessage(null);
  return {
    loading,
    setLoading,
    errorMessage,
    setErrorMessage,
  };
};

export const uploader = () => {
  const [textInputValue, setTextInputValue] = module.state.textInputValue('');
  return {
    textInputValue,
    setTextInputValue,
  };
};

export const postUploadRedirect = (storeState) => {
  const learningContextId = selectors.app.learningContextId(storeState);
  const blockId = selectors.app.blockId(storeState);
  return (videoUrl) => navigateTo(`/course/${learningContextId}/editor/video/${blockId}?selectedVideoUrl=${videoUrl}`);
};

export const onVideoUpload = () => {
  const storeState = store.getState();
  return module.postUploadRedirect(storeState);
};

const getFileExtension = (filename) => filename.slice(Math.abs(filename.lastIndexOf('.') - 1) + 2);

export const fileValidator = (setLoading, setErrorMessage, uploadVideo) => (file) => {
  const supportedFormats = Object.keys(extToMime);
  const ext = getFileExtension(file.name);
  const type = extToMime[ext] || '';
  const newFile = new File([file], file.name, { type });

  if (supportedFormats.includes(ext)) {
    uploadVideo({
      supportedFiles: [newFile],
      setLoadSpinner: setLoading,
      postUploadRedirect: onVideoUpload(),
    });
  } else {
    const errorMsg = 'Video must be an MP4 or MOV file';
    setErrorMessage(errorMsg);
  }
};

export default {
  postUploadRedirect,
  uploadEditor,
  uploader,
  onVideoUpload,
  fileValidator,
};
