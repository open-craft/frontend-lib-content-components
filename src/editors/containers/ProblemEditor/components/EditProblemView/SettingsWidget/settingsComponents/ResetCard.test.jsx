import React from 'react';
import { shallow } from 'enzyme';
import { formatMessage } from '../../../../../../../testUtils';
import { ResetCard } from './ResetCard';
import { resetCardHooks } from '../hooks';
import { useDispatch } from 'react-redux';

jest.mock('../hooks', () => ({
  resetCardHooks: jest.fn(),
}));


describe('ResetCard', () => {

  const props = {
    showResetButton: false,
    intl: { formatMessage },
  };

  const resetCardHooksProps = {
    setResetTrue: jest.fn().mockName('resetCardHooks.setResetTrue'),
    setResetFalse: jest.fn().mockName('resetCardHooks.setResetFalse'),
  };

  resetCardHooks.mockReturnValue(resetCardHooksProps);

  describe('behavior', () => {
    it(' calls resetCardHooks when initialized', () => {
        shallow(<ResetCard {...props} />);
        const dispatch = useDispatch();
        expect(resetCardHooks).toHaveBeenCalledWith( dispatch );
    });
});

  describe('snapshot', () => {
    test('snapshot: renders reset true setting card', () => {
      expect(shallow(<ResetCard {...props} />)).toMatchSnapshot();
    });
    test('snapshot: renders reset true setting card', () => {
      expect(shallow(<ResetCard {...props} showResetButton={true} />)).toMatchSnapshot();
    });
  });
});