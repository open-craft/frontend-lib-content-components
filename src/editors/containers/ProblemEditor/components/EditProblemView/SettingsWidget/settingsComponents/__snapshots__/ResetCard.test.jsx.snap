// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetCard snapshot snapshot: renders reset true setting card 1`] = `
<SettingsOption
  summary="False"
  title="Show reset option"
>
  <div
    className="halfSpacedMessage"
  >
    <span>
      <FormattedMessage
        defaultMessage="Determines whether a 'Reset' button is shown so the user may reset their answer, generally for use in practice or formative assessments."
        description="Reset settings card text"
        id="authoring.problemeditor.settings.reset.text"
      />
    </span>
  </div>
  <div
    className="spacedMessage"
  >
    <Hyperlink
      destination="#"
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Set a default value in advanced settings"
        description="Advanced settings link text"
        id="authoring.problemeditor.settings.advancedSettingLink.text"
      />
    </Hyperlink>
  </div>
  <ButtonGroup
    className="mb-2"
    size="lg"
  >
    <Button
      onClick={[MockFunction resetCardHooks.setResetFalse]}
      variant="primary"
    >
      <FormattedMessage
        defaultMessage="False"
        description="False option for reset"
        id="authoring.problemeditor.settings.reset.false"
      />
    </Button>
    <Button
      onClick={[MockFunction resetCardHooks.setResetTrue]}
      variant="outline-primary"
    >
      <FormattedMessage
        defaultMessage="True"
        description="True option for reset"
        id="authoring.problemeditor.settings.reset.true"
      />
    </Button>
  </ButtonGroup>
</SettingsOption>
`;

exports[`ResetCard snapshot snapshot: renders reset true setting card 2`] = `
<SettingsOption
  summary="True"
  title="Show reset option"
>
  <div
    className="halfSpacedMessage"
  >
    <span>
      <FormattedMessage
        defaultMessage="Determines whether a 'Reset' button is shown so the user may reset their answer, generally for use in practice or formative assessments."
        description="Reset settings card text"
        id="authoring.problemeditor.settings.reset.text"
      />
    </span>
  </div>
  <div
    className="spacedMessage"
  >
    <Hyperlink
      destination="#"
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Set a default value in advanced settings"
        description="Advanced settings link text"
        id="authoring.problemeditor.settings.advancedSettingLink.text"
      />
    </Hyperlink>
  </div>
  <ButtonGroup
    className="mb-2"
    size="lg"
  >
    <Button
      onClick={[MockFunction resetCardHooks.setResetFalse]}
      variant="outline-primary"
    >
      <FormattedMessage
        defaultMessage="False"
        description="False option for reset"
        id="authoring.problemeditor.settings.reset.false"
      />
    </Button>
    <Button
      onClick={[MockFunction resetCardHooks.setResetTrue]}
      variant="primary"
    >
      <FormattedMessage
        defaultMessage="True"
        description="True option for reset"
        id="authoring.problemeditor.settings.reset.true"
      />
    </Button>
  </ButtonGroup>
</SettingsOption>
`;
